### PR TITLE
Expose StreamingHandle on TokenStream

### PR DESCRIPTION
**Pull request by DragonFSKY**
## Issue
Part of #5275

## Change
This PR adds an opt-in `TokenStream.onStreamingHandle(...)` callback so adapters can observe the existing `StreamingHandle` without changing the current plain token callback path.

The callback is a default no-op, so existing `TokenStream` implementations keep their current behavior unless they choose to expose a handle.

Changes included:
- add `TokenStream.onStreamingHandle(...)`;
- support the callback in `AiServiceTokenStream`;
- observe the handle in `AiServiceStreamingResponseHandler` from context-bearing partial response, thinking, and tool-call callbacks;
- preserve the existing `onPartialResponse(...)` token path;
- add tests for handle observation, duplicate callback registration, deduplication, and thinking/tool-call context propagation.

This is intended to support a companion `langchain4j-spring` PR that propagates Reactor `Flux` cancellation to `StreamingHandle.cancel()`.

## Verification
- [x] `./mvnw -pl langchain4j -Dtest=AiServiceTokenStreamStreamingHandleTest,AiServiceTokenStreamTest#start_onStreamingHandleInvokedMultipleTimes_shouldThrowException test`
- [x] `./mvnw -pl langchain4j -DskipTests install`
- [x] `mvn -Pspotless validate`
- [x] `git diff --check origin/main...HEAD`

## General checklist
- [X] There are no breaking changes (API, behaviour)
- [X] I have added unit and/or integration tests for my change
- [X] The tests cover both positive and negative cases
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new maven module
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`

## Checklist for adding new embedding store integration
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

File: langchain4j/src/main/java/dev/langchain4j/service/AiServiceStreamingResponseHandler.java [hunk 1]
```diff
@@ -413,6 +423,11 @@ public void onCompleteResponse(ChatResponse chatResponse) {
                     // completing
                     if (partialResponseHandler != null) {
                         responseBuffer.forEach(partialResponseHandler::accept);
+                    } else if (partialResponseWithContextHandler != null) {
+                        PartialResponseContext replayContext =
+                                new PartialResponseContext(new CancellationUnsupportedStreamingHandle());
```
Comment (dliubarskyi) [inline langchain4j/src/main/java/dev/langchain4j/service/AiServiceStreamingResponseHandler.java]: can't we try using `observedStreamingHandle` instead of `CancellationUnsupportedStreamingHandle` (if available)?

File: langchain4j/src/main/java/dev/langchain4j/service/AiServiceStreamingResponseHandler.java [hunk 2]
```diff
@@ -99,6 +101,8 @@ class AiServiceStreamingResponseHandler implements StreamingChatResponseHandler
     private final List<String> responseBuffer = new ArrayList<>();
     private final boolean hasOutputGuardrails;
 
+    private StreamingHandle observedStreamingHandle;
```
Comment (dliubarskyi) [inline langchain4j/src/main/java/dev/langchain4j/service/AiServiceStreamingResponseHandler.java]: Shouldn't it be `volatile` or `AtomicReference`? `onPartialResponse` and other callbacks could be called from different threads

**Comment by DragonFSKY** [comment]
Thanks for the review. I rebased this on the latest main and addressed both points:

- `observedStreamingHandle` now uses `AtomicReference` so callbacks from different threads have safe visibility.
- Context replay and legacy callbacks now reuse the observed streaming handle when available, falling back to `CancellationUnsupportedStreamingHandle` only when no handle has been observed.

I also added a regression test covering the mixed context/no-context partial response path.

**Comment by DragonFSKY** [comment]
Before I continue with the companion Spring PR, could you confirm whether this `TokenStreamAdapter`-based approach is still the preferred direction given the `PublisherAdapter` work in #5527? If #5527 supersedes this path, I’m happy to close both PRs. I’ll keep the companion PR in draft in the meantime.
